Implement SMTP StartTLS and Adaptive Auth

james-d-elliott opened the pull request:
- If the StartTLS extension is advertised we automatically StartTLS
- Uses the secure config key to determine if we should verify the cert
- Attempt auth when the config has a SMTP password and the server supports the AUTH extension
- Check the mechanisms supported by the server and use PLAIN, LOGIN, or CRAM-MD5 depending on which is supported
- Changed secure key to use boolean values instead of strings
- Arranged SMTP notifier properties/vars to be in the same order

Fixes #464 

I've built and published an image on docker hub: [jamesdelliott/authelia:smtp-adaptive-auth](https://hub.docker.com/layers/jamesdelliott/authelia/smtp-adaptive-auth/images/sha256-d07bc0e6a4068d398f9d2b8d905d251fe062a07d4fd244cf857a180fb3bfd4d0)

Requires Testing:
- [x] SMTP Server that doesn't Support StartTLS (should succeed)
- [x] SMTP Server that does support StartTLS (should succeed)
- [x] SMTP Server that doesn't support AUTH (should succeed)
- [x] SMTP Server that only supports AUTH CRAM-MD5 (should succeed)
- [x] SMTP Server that only supports AUTH PLAIN (should succeed)
- [x] SMTP Server that only supports AUTH NTLM (should fail with error)
- [x] SMTP Server that only supports AUTH LOGIN (should succeed)
- [x] SMTP Server that supports AUTH PLAIN and AUTH LOGIN (should use AUTH PLAIN)
- [x] SMTP Server with bad cert and secure set to False (should succeed)
- [x] SMTP Server with bad cert and secure set to True (should fail)
- [x] SMTP Server with bad cert and secure commented out, or unset (should succeed)
- [x] SMTP Server with good cert and secure set to true (should succeed)
- [x] SMTP Server with good cert and secure commented out, or unset (should succeed, wont verify cert)
- [x] SMTP Server that only offers AUTH LOGIN after STARTTLS

Things I plan to add to this PR:
- [x] Debug Logging for STARTTLS
- [x] Debug Logging for AUTH
- [x] Rebase / Squash
- [x] Finalize Spec


## Final Specification Tests

Tests unrelated to password/auth (should be possible in test suite):
- [x] Server that doesn't support STARTLS (should succeed)
- [x] Server that does support STARTLS (should succeed)
- [x] Server with self-signed cert and secure set to False (should succeed)
- [x] Server with self-signed cert and secure set to True (should fail)
- [x] Server with self-signed cert and secure commented/unset (should succeed)
- [x] Server with good cert and secure set to True (should succeed)
- [x] Server with good cert and secure commented/unset (should succeed)

Tests WITH password set in config (should be possible in test suite):
- [x] Server that doesn't support AUTH (should fail)
- [x] Server that only supports an AUTH mechanism NOT supported by Authelia (should fail)
- [x] Server that only supports AUTH PLAIN (should succeed)
- [x] Server that only supports AUTH LOGIN (should succeed)
- [x] Server that only offers AUTH LOGIN after STARTTLS (should succeed)

Tests WITHOUT password set in config (should be possible in test suite):
- [x] Server that doesn't support AUTH (should succeed)
- [x] Server that has optional AUTH (should succeed)
- [x] Server that has required AUTH (should fail)

Manual tests (probably not possible in test suite):
- [x] Server that supports AUTH PLAIN and AUTH LOGIN (should use AUTH PLAIN)

james-d-elliott commented:
Question. In a situation where the config defines a password, but the server does not support AUTH, do we want to throw an error or just skip AUTH and attempt to send the mail?

james-d-elliott commented:
Not sure how to debug the CI errors. Though nothing changed should cause issues with that suite.

nightah commented:
> Not sure how to debug the CI errors. Though nothing changed should cause issues with that suite. The fact the original commit passed the failed test suggests it was a false negative.

The issue was unrelated and seems to be a problem with Travis, I'll kick it back off in a second.

clems4ever commented:
Thanks @james-d-elliott for the contribution.

> Questions.
> 
> 1. In a situation where the config defines a password, but the server does not support AUTH, do we want to throw an error or just skip AUTH and attempt to send the mail like this: https://github.com/james-d-elliott/authelia/blob/feature/notification-smtp-adaptive-auth-and-starttls/internal/notification/smtp_notifier.go#L99-L101

I think we can throw an error. It might help to capture a misconfiguration in a SMTP server set up for Authelia.

> 2. Also in a situation where the SMTP server does not advertise auth mechanisms that Authelia supports do we want to skip AUTH or just return an error like this: https://github.com/james-d-elliott/authelia/blob/feature/notification-smtp-adaptive-auth-and-starttls/internal/notification/smtp_notifier.go#L90-L92

I was a bit concerned about the automatic upgrade from no auth to AUTH because I always wonder if the upgrade has truly been done. I don't really like what is implicit. I'd be glad to have an error but this means introducing a new configuration flag then. What do you think?

clems4ever commented:
Do you guys know how to configure a SMTP service in all those cases to introduce those use cases as integration tests? That would ensure non-regression for the future.
For instance, do you know how to configure postfix in all those cases?

nightah commented:
> Do you guys know how to configure a SMTP service in all those cases to introduce those use cases as integration tests? That would ensure non-regression for the future.
> For instance, do you know how to configure postfix in all those cases?

I can help with this part, NTLM obviously will require AD/Exchange but the other cases we can handle.

james-d-elliott commented:
I think for point 1 and 2 an error should be thrown. If authentication is configured you expect the server to support it. The way I've done STARTTLS also avoids the situation of servers that only offer auth mechanisms after STARTTLS. I think it's sufficient to set both 1 and 2 as failure conditions and is probably preferable, also I think that's what you're suggesting. 

As far as setup of a postfix server it should be easy. With the 4 auth mechanisms implemented it covers most configurations (anonymous, PLAIN, CRAM-MD5, LOGIN). I only tested NTLM to verify the expected behavior of specifically when an unsupported mech was the only one available.

james-d-elliott commented:
Some fake smtp servers to look at:

https://github.com/kirviq/dumbster
https://www.npmjs.com/package/fake-smtp-server

james-d-elliott commented:
Unless there are objections, I think this would provide the most portable and lightweight method of testing: https://nodemailer.com/extras/smtp-server/

nightah commented:
> Also some points on the decision of the order of auth mechanism preference. PLAIN is supported and maintained by the golang library and we know it works, LOGIN is custom, and CRAM-MD5 is untested.

To be honest I've never seen CRAM-MD5 out in the wild, I think my preference would be to just support PLAIN and LOGIN as I do not believe there's an explicit case to support CRAM-MD5.

I don't have an issue with the suggested method for testing, I believe the only reason @clems4ever suggested Postfix was because you're likely to run in to a lot of mail servers/relays that utilise Postfix.

nightah commented:
https://hub.docker.com/r/catatnight/postfix while not updated could even be used for the testing.

@clems4ever if you still want to stick to postfix and provide our own configuration we could simply just do something very similar to this: https://github.com/catatnight/docker-postfix/blob/master/assets/install.sh

james-d-elliott commented:
I'm fine with ditching CRAM-MD5, I've not seen many services that don't use PLAIN or LOGIN (or both). Even MS Exchange supports LOGIN after STARTTLS, and NTLM is hardly worth testing at all - I just used it to test a failure (unknown AUTH mech). I mainly implemented it because it was part of the core library. I also have no real preference on test suite direction, though I'm willing to help once one is chosen.

If we're wanting to test STARTTLS as part of the test suite with Secure = True, it will probably be complicated - it may be worth skipping. We would either need to add certs to the CI/CD test VM's trust store or we'll have to do something with the tlsconfig (crypto/tls.Config).

nightah commented:
> If we're wanting to test STARTTLS as part of the test suite with Secure = True, it will probably be complicated - it may be worth skipping. We would either need to add certs to the CI/CD test VM's trust store or we'll have to do something with the tlsconfig (crypto/tls.Config).

We can include the certs in the `internal/suites/x` path along with other configuration relevant to the new suite. Or we could just generate a self-signed cert alongside the service.

james-d-elliott commented:
I will probably have the following done in the next 24 hours:

1. [x] Moving internal/utils/smtp_login_auth.go to internal/notifier/smtp_login_auth.go
2. [x] Throwing an error in both of the following circumstances if the user configured a password:
 a. Server doesn't advertise AUTH extension.
 b. Server doesn't advertise AUTH mechanism that Authelia supports.
3. [ ] Rebasing with authelia/authelia:master so the merge is easy.
4. [ ] A retest of all functionality.

I'm pretty sure that will make it merge/release ready, though I'm not sure if you want a test suite before merging or not. If you do I may need some assistance with that. I will at least make a list of things to test.

@clems4ever can you confirm the following:

1. In a situation where smtp password is set, we want to force authentication. So if the server specified does not offer AUTH or the mechanisms it offers are not compatible the sending of the email in this situation will fail (user should configure correctly).
2. Do we want to leave the CRAM-MD5 mechanism in there. I have tested it and it works as I expected without changes. It's not hugely popular as mentioned by @nightah but since it works it may be worth keeping - though it may be harder to test than LOGIN or PLAIN, and most SMTP servers offer at least one of these (ms-exchange even does after STARTTLS).

clems4ever commented:
Great job @james-d-elliott ! And thank you @nightah for the support.

I'm ok with 1. and 2. I'm also ok to merge the feature when it's ready and we'll add the integration tests in another PR. I'll show you how to do it so that next time you're autonomous on that part.

Regarding the solution to test the feature, I'd rather use postfix if you guys can help me build the various configurations to run the tests against (I'm a noob in postfix config), otherwise the nodemailer solution would fit too. I did not know that they had a complete SMTP server implementation. So to summarize, I'd prefer postfix because it would also provide examples on how to setup a SMTP server in a target environment but I need assistance to have the configs running in a docker container otherwise I'll fallback to nodemailer.

For certs, if we don't test with verify=true, that's fine for the beginning. It's a point which has been skipped in other tests as well anyway.

@james-d-elliott , let us know when we can have a final review of your work before merging.

clems4ever commented on internal/notification/smtp_notifier.go:
```diff
@@ -3,50 +3,106 @@ package notification
+				log.Debug("SMTP server supports AUTH CRAM-MD5, attempting")
+			}
+			if auth == nil {
+				return errors.New("SMTP server does not advertise a AUTH mechanism that Authelia Supports")
```
Idea for the future: we should probably introduce a hook at Authelia startup to perform early checks of configuration: "is SMTP server correctly configured?", "is db correctly configured?", "are credentials against service X ok?"
This way users could notice failures early during set up.

james-d-elliott replied:
This is very doable, and would probably force me to get more familiar with the code base.

james-d-elliott replied:
I will look into this. As you said future I think for the purposes of the PR you intend to resolve it prior to merging; but I'll leave this open for you to decide.

This part will probably involve two things, implement each stage of the send mail process into a function which returns an error or nil (i.e. STARTTLS, AUTH). I'm going to quickly see if I can implement this split of functionality now without breaking anything.

james-d-elliott replied:
I had a look at this, but I think it may require a bit more thought and full implementation instead of just splitting the functionality.

clems4ever replied:
Definitely not a quick win, this is why I mentioned "for the future" meaning later in another review. The PR as it is is fine for me, the goal is reached. Thanks.

clems4ever commented on internal/notification/smtp_notifier.go:
```diff
@@ -3,50 +3,106 @@ package notification
+				log.Debug("SMTP AUTH completed without error")
+			}
+		} else {
+			log.Debug("SMTP config has a password provided but server does not support AUTH, skipping")
```
As suggested, there should be an error thrown here because user provides a password which is not consumed by the server.

james-d-elliott replied:
Done.

clems4ever commented on internal/utils/smtp.go:
```diff
@@ -0,0 +1,34 @@
+package utils
```
Please move this file to the notifier package as suggested.

james-d-elliott replied:
Done.

james-d-elliott commented:
I've rebased and squashed everything. I only have to create a new list of tests, test it all, and ensure there are no obvious mistakes. You're fine to do other commits obviously.

james-d-elliott commented:
I believe this is ready for a merge. I have added a full list of tests I believe are suitable for these changes, tested them, and had a thorough read of the code.

I identified one issue and have force pushed a fix to it. I put in a check before AUTH PLAIN to check if STARTTLS occurred before attempting AUTH, this was for testing for possible future functionality which was not intended for my PR.

clems4ever commented on internal/notification/smtp_notifier.go:
```diff
@@ -1,52 +1,112 @@
+
+	// Attempt AUTH if password is specified only
+	if n.password != "" {
+
```
Add a warning when authentication takes place without STARTTLS. This should raise a security red flag to the administrator.
We could print something like "Authentication occurred over an insecure connection. Using STARTTLS is recommended to keep credentials safe."

james-d-elliott replied:
Good idea.

james-d-elliott commented:
> .

This is all done. I plan to keep helping to improve it. The code base is very understandable and well thought out, and I really like Authelia. I have a working startup check for reference, but not overly happy with it. I'll probably make a separate draft PR so you can take a look.

I'm curious what you mean about full TLS support, do you mean SMTPS instead of STARTTLS? Or some additional configuration?

nightah commented:
SMTPS has more or less been deprecated for a long time now. STARTTLS is the de-facto default in the mailserver world for default port 25 and submission typically supports STARTTLS too.

Perhaps @clems4ever means TLS across all of Authelia including the webserver and ldap implementations?

clems4ever commented:
> I'm curious what you mean about full TLS support, do you mean SMTPS instead of STARTTLS? Or some additional configuration?

I indeed meant connection over TLS from beginning to end. Not only an upgrade for credentials exchange but I did not know that SMTPS has been deprecated though and I misunderstood the STARTTLS command. I was thinking it was only upgrading to TLS for credential exchange but it actually upgrade the entire connection so that's all fine, you can forget what I wrote. Sorry.

clems4ever commented:
> This is all done. I plan to keep helping to improve it. The code base is very understandable and well thought out, and I really like Authelia. I have a working startup check for reference, but not overly happy with it. I'll probably make a separate draft PR so you can take a look.

Very nice! Regarding your attempt I think you should isolate the checks in a method of the configuration package so that it's easier to unit test in isolation. The aim is to have multiple kind of checks done eventually (connection tests, user hashes validity checks, ldap binding checks, etc).

clems4ever commented:
Also regarding this PR, since we won't support SMTPS which is totally acceptable given it has been deprecated, I think we should have an option to force STARTTLS to avoid downgrade attacks.

clems4ever commented:
I've created an issue for what I propose: https://github.com/authelia/authelia/issues/524.
Let me know what you think about it.
